### PR TITLE
[EC-206] - Move functions from core to specific domain

### PR DESCRIPTION
**Pull request by christian-calabrese**
### List of changes
Move functions from core to specific domain
- function_admin
- function_app_async
- function_app
- function_assets_cdn
- function_public
- function_services
- function_shared

### Motivation and context

With the aim of reducing plan time and reorganize the terraform code as described in this [RFC](https://pagopa.atlassian.net/wiki/spaces/ENG/pages/910492147/RFC-ENG+Riorganizzazione+configurazioni+Terraform), resources are being subdivided in domains.

To reduce the chances of error, resources are being moved without applying module upgrades, which will be executed right after the move with a new PR.

### Type of changes

- [ ] Add new resources
- [X] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [X] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [X] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [X] No

### Other information

---

### How to apply

After PR is approved
1. run deploy pipeline from Azure DevOps [io-platform-iac-projects](https://dev.azure.com/pagopaspa/io-platform-iac-projects/_build?view=folders)
2. select PR branch
3. wait for approval

### CODE REVIEW

**Comment by gunzip** [comment]
Try to keep PR in draft until they pass CI ;-)

File: .github/workflows/functions_ci.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  AZURE_ENVIRONMENT: prod
+
+jobs:
+
+  terraform_plan_job:
+    name: Terraform Plan
+    runs-on: self-hosted
+    environment: prod-ci
+    env:
+      ARM_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
```
Comment (Krusty93) [inline .github/workflows/functions_ci.yml:31]:
```suggestion
      ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
```

File: .github/workflows/functions_cd.yml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  ARM_STORAGE_USE_AZUREAD: true
+
+jobs:
+
+  terraform_plan_job:
+    name: Terraform Plan
+    runs-on: self-hosted
+    environment: prod-ci
+    env:
+      ARM_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
```
Comment (Krusty93) [inline .github/workflows/functions_cd.yml:29]:
```suggestion
      ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
```

File: .github/workflows/functions_cd.yml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          bash ./terraform.sh plan ${{ env.AZURE_ENVIRONMENT }} -lock-timeout=3000s -out=tfplan-${{ env.AZURE_ENVIRONMENT }}-${{ github.sha }} -input=false
+
+      - uses: actions/upload-artifact@c7d193f32edcb7bfad88892161225aeda64e9392
+        id: artifact_upload
+        name: Upload plans as artifacts
+        with:
+          name: tfplan-output
+          if-no-files-found: error
+          path: |
+            **/tfplan-prod-*
```
Comment (Krusty93) [inline .github/workflows/functions_cd.yml:77]:
```suggestion
            **/tfplan-${{ env.AZURE_ENVIRONMENT }}-*
```

File: .github/workflows/functions_cd.yml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    outputs:
+      terraform_version: ${{ steps.env_tf_version.outputs.terraform_version}}
+
+  terraform_apply_job:
+    name: Terraform Apply
+    runs-on: self-hosted
+    environment: prod-cd
+    needs: [terraform_plan_job]
+    env:
+      ARM_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
```
Comment (Krusty93) [inline .github/workflows/functions_cd.yml:88]:
```suggestion
      ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
```

File: src/domains/functions/test_users.tf [hunk 5]
```diff
@@ -0,0 +1,274 @@
+locals {
```
Comment (Krusty93) [inline src/domains/functions/test_users.tf:1]: We already have this file as module under `domains`, you can reference it

**Review by Krusty93 — APPROVED** [review]

**Comment by gunzip** [comment]
well done @christian-calabrese !

File: .github/workflows/functions_cd.yml [hunk 6]
```diff
@@ -0,0 +1,138 @@
+name: Continuous Delivery on prod functions
```
Comment (gunzip) [inline .github/workflows/functions_cd.yml:1]: From the name, one can deduce that this action deploys the function while applying infrastructure changes. Better name: "Appy Terraform changes for domain functions" or something alike.

File: .github/workflows/functions_ci.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
+name: Continuous Integration on prod functions
```
Comment (gunzip) [inline .github/workflows/functions_ci.yml:1]: Ditto.

**Pull request merged**